nym-api/src/coconut/helpers.rs

```diff
@@ -50,15 +62,120 @@ impl CredentialRequest for FreePassRequest {
 
 pub(crate) fn blind_sign<C: CredentialRequest>(
     request: &C,
-    signing_key: &SecretKey,
+    signing_key: &SecretKeyAuth,
 ) -> Result<BlindedSignature, CoconutError> {
-    let public_attributes = request.public_attributes();
-    let attributes_ref = public_attributes.iter().collect::<Vec<_>>();
-
-    Ok(nym_coconut::blind_sign(
-        bandwidth_credential_params(),
-        signing_key,
-        request.blind_sign_request(),
-        &attributes_ref,
+    Ok(nym_compact_ecash::scheme::withdrawal::issue(
+        bandwidth_credential_params().grp(),
+        signing_key.clone(),
+        request.ecash_pubkey().clone(),
+        request.withdrawal_request(),
+        request.expiration_date(),
     )?)
 }
+
+pub(crate) struct CoinIndexSignatureCache {
+    pub(crate) epoch_id: AtomicU64,
+    pub(crate) signatures: RwLock<Option<Vec<CoinIndexSignature>>>,
+}
+
+impl CoinIndexSignatureCache {
+    pub(crate) fn new() -> Self {
+        CoinIndexSignatureCache {
+            epoch_id: AtomicU64::new(u64::MAX),
+            signatures: RwLock::new(None),
+        }
+    }
+    // if the epoch id cached is the one expected, return the cached signatures, else return None
+    pub(crate) async fn get_signatures(
+        &self,
+        expected_epoch_id: u64,
+    ) -> Option<Vec<CoinIndexSignature>> {
+        if self.epoch_id.load(Ordering::Acquire) == expected_epoch_id {
+            let signatures = self.signatures.read().await;
+            signatures.clone()
+        } else {
+            None
+        }
+    }
+
+    // refreshes (if needed) and returns the signatures.
+    pub(crate) async fn refresh_signatures(
+        &self,
+        expected_epoch_id: u64,
+        ecash_parameters: &Parameters,
+        verification_key: &VerificationKeyAuth,
+        secret_key: &SecretKeyAuth,
+    ) -> Vec<CoinIndexSignature> {
+        let mut signatures = self.signatures.write().await;
+
+        //if this fails, it means someone else updated the signatures in the meantime
+        // => We don't have to update them, and we know they exist
+        // (this check can spare us some signing)
+        if self.epoch_id.load(Ordering::Acquire) != expected_epoch_id {
+            *signatures = Some(sign_coin_indices(
+                ecash_parameters,
+                verification_key,
+                secret_key,
+            ));
+            self.epoch_id.store(expected_epoch_id, Ordering::Release);
+        }
+
+        signatures.clone().unwrap() // Either we or someone else update the signatures, so they must be there
```

Comment:
`signatures` is `None` only at startup, to avoid useless computations. The first call to that `fn` WILL set `signatures`, and it is never put to `None` anywhere after that, so there will be something

Reply:
Won't fix